[3756] Support component instance drag & drop (i.e. browse)

aachavarria opened the pull request:
https://github.com/craftercms/craftercms/issues/3756

aachavarria committed:
BrowsePanel Init

rart reviewed (changes requested):
Let’s call the component BrowseComponentsPanel. Rename the panel id all over the place too.
I take it you’re just getting started but we need to clean out and refactor the repetition of styles and other code with ComponentsPanel

aachavarria committed:
renaming to browsecomponentspanel,
adding actions to fetch contenttypecomponents

aachavarria committed:
Merge branch 'feature/3632' of https://github.com/yureii00/studio-ui into feature/3756

 Conflicts:
	ui/app/src/state/actions/preview.ts
	ui/app/src/state/reducers/preview.ts

aachavarria committed:
creating epic, reducer, actions for fetchContentTypeComponents

aachavarria committed:
Merge branch '2019-11-ui-restructure' of https://github.com/craftercms/studio-ui into feature/3756

aachavarria committed:
Merge branch '2019-11-ui-restructure' of https://github.com/craftercms/studio-ui into feature/3756

aachavarria committed:
creating new interface 'SearchContentInstance',
renaming actions fetchContentTypeComponents to fetchComponentsByContentType,
creating getContentInstance

aachavarria committed:
AssetsPanel: removing unused translation,
BrowseComponentPanel: adding pagination, search, component,
Component: moving component to a export component,
ComponentsPanel: moving component to a export component,
preview.ts: adding partial<ComponentsContentTypeParams> to fetchComponentsByContentType

aachavarria committed:
renaming Component to PanelListItem

aachavarria committed:
adding Select dropdown for contentTypes

aachavarria committed:
searchBar: adding disabled.
AssetsPanel: shouldResolve changes.
BrowseComponentsPanel: Adding contenttype dropdown.
PreviewConfierge: adding contentTypeFilter conditional.
Content.ts: filtering getContentByContentType() by type components

aachavarria committed:
removing console.log,
adding and removing effect dependencies

aachavarria committed:
basic dnd for browseComponents

aachavarria committed:
BrowseComponentsPanel: onDragStart using instance.
ComponentesPanel: Adding browseSharedInstance on menu click 'Browse "shared" instances'.
PreviewConcierge: adding insertInstance.
Content.ts: adding insertInstance.
Action.Preview.ts: adding browseSharedInstance action, and constants.
Reducer.Preview.ts: adding BROWSE_SHARED_INSTANCE reducer.
ContentController.js: adding insertInstance.
Guest.js: adding onHostInstanceDragStarted/onHostInstanceDragEnd.
GuestProxy.js: adding INSERT_INSTANCE_OPERATION,
Utils.js: adding stripEscapedHtmlTags and decodeHTML

aachavarria committed:
browseComponentsPanel fix, not loading new components on filterChange

aachavarria committed:
adding validation

aachavarria committed:
revert commit

aachavarria committed:
adding registerDeregisterElements function for 'move' and 'insert' actions

rart commented on ui/app/src/models/PreviewToolIDs.ts:
```diff
@@ -21,6 +21,7 @@ type Tools =
   'craftercms.ice.ice' |
-  'craftercms.ice.editForm';
+  'craftercms.ice.editForm' |
+  'craftercms.ice.browse';
```
browseComponents

rart commented on ui/app/src/modules/Preview/Tools/BrowseComponentsPanel.tsx:
```diff
@@ -0,0 +1,328 @@
+      dispatch(fetchComponentsByContentType(null, { keywords }));
+    });
+    return () => subscription.unsubscribe();
+  }, [dispatch, onSearch$]);
```
Looks like we're using this in a couple of places. Let's refactor on to a reusable custom hook and update the different places to use.

```js
// @BrowseComponentsPanel.tsx, @AssetsPanel.tsx, etc
const onSearch = useMemo(() => (
  (keywords: string) => dispatch(fetchComponentsByContentType(null, { keywords }))
), [dispatch]);

const onSearch$ = useDebouncedInput(onSearch);

// @hooks.ts
export function useDebouncedInput(observer: (keywords: string) => any) {
  const subject$Ref = useRef(new Subject<string>());
  useEffect(() => {
    const subscription = subject$Ref.current.pipe(
      debounceTime(400),
      distinctUntilChanged()
    ).subscribe(observer);
    return () => subscription.unsubscribe();
  }, [observer]);
  return subject$Ref.current;
}
```

rart commented on ui/app/src/modules/Preview/Tools/BrowseComponentsPanel.tsx:
```diff
@@ -0,0 +1,328 @@
+            onPageChanged={onPageChanged}
+            onDragStart={onDragStart}
+            onDragEnd={onDragEnd}
+            AUTHORING_BASE={AUTHORING_BASE}
```
Looks awkward a prop like this... let's just use camelcase: `authoringBase={AUTHORING_BASE}`

rart commented on ui/app/src/modules/Preview/Tools/BrowseComponentsPanel.tsx:
```diff
@@ -0,0 +1,328 @@
+        ) : (
+          <EmptyState
+            title={formatMessage(translations.chooseContentType)}
+            image={`${AUTHORING_BASE}/static-assets/images/choose_option.svg`}
```
Switch to `chooseContentTypeImageUrl` prop. Then we won't need the AUTHORING_BASE at all.

rart commented on ui/app/src/modules/Preview/Tools/ComponentsPanel.tsx:
```diff
@@ -184,7 +172,7 @@ export function ComponentsPanelUI(props) {
       >
         <MenuItem onClick={onMenuOptionClicked}>List in-page instances</MenuItem>
-        <MenuItem onClick={onMenuOptionClicked}>Browse "shared" instances</MenuItem>
+        <MenuItem onClick={onBrowseSharedInstancesClicked}>Browse "shared" instances</MenuItem>
```
Translate menu item option. Change to "Browse existing instances".

rart commented on ui/app/src/modules/Preview/Tools/PanelListItem.tsx:
```diff
@@ -0,0 +1,98 @@
+  onMenu?: (anchor: Element) => any;
+}
+
+export function PanelListItem(props: PanelListItemProps) {
```
Let's name "DraggablePanelListItem"

rart commented on ui/app/src/services/content.ts:
```diff
@@ -49,6 +49,54 @@ export function getDOM(site: string, path: string): Observable<XMLDocument> {
+  'createdDate_dt',
+  'lastModifiedDate',
+  'lastModifiedDate_dt'
+];
```
Need to review this together.

rart commented on ui/app/src/services/content.ts:
```diff
@@ -562,6 +610,74 @@ export function insertComponent(
   );
 }
 
+export function insertInstance(
```
Should be using the performOperation function. Looks like it doesn't consider embedded.

aachavarria replied:
we can see this on call

rart commented on ui/app/src/state/actions/preview.ts:
```diff
@@ -77,6 +81,10 @@ export const CHANGE_CURRENT_URL = 'CHANGE_CURRENT_URL';
+export const FETCH_COMPONENTS_BY_CONTENT_TYPE = 'FETCH_COMPONENTS_BY_CONTENT_TYPE';
+export const FETCH_COMPONENTS_BY_CONTENT_TYPE_COMPLETE = 'FETCH_COMPONENTS_BY_CONTENT_TYPE_COMPLETE';
+export const FETCH_COMPONENTS_BY_CONTENT_TYPE_FAILED = 'FETCH_COMPONENTS_BY_CONTENT_TYPE_FAILED';
+export const BROWSE_SHARED_INSTANCE = 'BROWSE_SHARED_INSTANCE';
```
This is a message that is shared between host/guest, correct? **If so**, move to the "Accommodation Actions" section above.

Rename "BROWSE_COMPONENT_INSTANCES"

rart commented on ui/app/src/state/actions/preview.ts:
```diff
@@ -187,4 +195,17 @@ export const fetchAssetsPanelItemsComplete = createAction<SearchResult>(FETCH_AS
 
 export const fetchAssetsPanelItemsFailed = createAction(FETCH_ASSETS_PANEL_ITEMS_FAILED);
 
+export function fetchComponentsByContentType(contentTypeFilter?: string, options?: Partial<ComponentsContentTypeParams>): StandardAction {
```
use createAction

aachavarria replied:
pending

rart commented on ui/app/src/state/reducers/preview.ts:
```diff
@@ -274,7 +289,44 @@ const reducer = createReducer<GlobalState['preview']>({
-  })
+  }),
+  [FETCH_COMPONENTS_BY_CONTENT_TYPE]: (state, { payload: { contentTypeFilter, options } }: { payload: { contentTypeFilter: string[] | string, options?: ComponentsContentTypeParams } }) => {
+    let new_query = { ...state.components.query, ...options };
```
camelCase

rart commented on ui/guest/src/classes/ContentController.js:
```diff
@@ -290,6 +291,40 @@ export class ContentController {
+    const model = models[modelId];
+    // A model with a field that hasn't been initialized with
+    // content may not have that field at all
+    let collection = ModelHelper.value(model, fieldId);
```
We need to consider nested here (and everywhere else). Need to use the ModelHelper.extractCollection. There's an example in one of the ops already. Update the other ops that need this too please.

Look for refactoring opportunities between the different operations.

rart commented on ui/guest/src/components/Guest.js:
```diff
@@ -471,6 +470,69 @@ export function Guest(props) {
 
     },
 
+    onHostInstanceDragStarted(instance) {
```
Look for refactoring opportunities (minimize duplicate code)

rart commented on ui/guest/src/components/GuestProxy.js:
```diff
@@ -62,6 +64,33 @@ export function GuestProxy(props) {
+        } else {
+          from = newIndex;
+          to = oldIndex + 1;
+        }
```
Looks good but double check the numbers are accurate as to how .slice works

rart commented on ui/guest/src/components/GuestProxy.js:
```diff
@@ -62,6 +64,33 @@ export function GuestProxy(props) {
 
     };
 
+    const registerDeregisterElements = (collection, type, newIndex, oldIndex) => {
```
registerDeregisterElements => updateElementRegistrations

rart commented on ui/guest/src/components/GuestProxy.js:
```diff
@@ -126,29 +155,7 @@ export function GuestProxy(props) {
-              }
-
-            });
+          registerDeregisterElements(Array.from($el.parent().children()), 'move', newIndex, index);
```
I think jQuery does give you an array, no?

aachavarria replied:
no, it returns a htmlCollection

rart commented on ui/guest/src/components/GuestProxy.js:
```diff
@@ -277,19 +284,39 @@ export function GuestProxy(props) {
-          forEach(
-            $daddy.children(),
-            (el, i) => {
+          registerDeregisterElements(Array.from($daddy.children()), 'insert', targetIndex);
```
I think jQuery does give you an array, no?

aachavarria replied:
it gives a htmlCollection and it does not have forEach

rart commented on ui/guest/src/components/GuestProxy.js:
```diff
@@ -277,19 +284,39 @@ export function GuestProxy(props) {
+                //cleaning CDATA and escapedhtmltags
+                let value = instance[field].replace("<![CDATA[", "").replace("]]>", "");
+                value = stripEscapedHtmlTags(value);
+                $clone.find(`[data-craftercms-field-id="${field}"]`).html(value);
```
You should not be getting CDATA or serialization details at this point. When parsing the content, the service should parse to "app format". Check service.

aachavarria replied:
for CDATA that is true, but for htmlTags there are needed because, we are retrieving the text inside of a RTE

rart commented on ui/guest/src/util.js:
```diff
@@ -523,3 +526,13 @@ export function removeLastPiece(str, splitChar = '.') {
+
+export function stripEscapedHtmlTags(str) {
+  return decodeHTML(str).replace(/<[^>]*>/g, '');
+}
```
Do we still need this? Provided that the service should give you clean ready-to-use data.

aachavarria replied:
yes, will talk about it

aachavarria committed:
renaming 'craftercms.ice.browse' to 'craftercms.ice.browseComponents'

aachavarria committed:
refactor of 'useDebouncedInput'

aachavarria committed:
PR review

aachavarria committed:
renaming PanelListItem' to 'DraggablePanelListItem'

aachavarria committed:
insertInstance: adding ParentModelId and PerformMutation.
content.ts: Refactoring `parseContentXML` WIP

rart commented on ui/app/src/modules/Preview/Tools/AssetsPanel.tsx:
```diff
@@ -53,10 +53,6 @@ const translations = defineMessages({
     id: 'craftercms.ice.assets.noResults',
     defaultMessage: ' No results found.'
   },
-  retrieveAssets: {
```
Why remove this one? This is the more appropriate loading text

aachavarria committed:
AssetsPanel: adding retrieveAssets;
BrowseComponentsPane: filtering components by contentType;
content.ts: refactor parseContentXML;
utils: adding new utils wrapElementInAuxDocument(), objectIdFromPath(), decodeHTML()

aachavarria committed:
refactor result from collection, to getResult()

rart commented on ui/app/src/services/content.ts:
```diff
@@ -562,6 +646,69 @@ export function insertComponent(
+        path: modelId,
+        unlock: 'true',
+        fileName: getInnerHtml(doc.querySelector('file-name'))
+      };
```
x

rart commented on ui/app/src/services/content.ts:
```diff
@@ -562,6 +646,69 @@ export function insertComponent(
+      return post(
+        writeContentUrl(qs),
+        serialize(doc)
+      );
```
The mutation function should only perform the necessary document mutations. Shouldn't do any sort of serializing, writing, etc.

rart commented on ui/app/src/services/content.ts:
```diff
@@ -562,6 +646,69 @@ export function insertComponent(
+        ...{
+          include: path,
+          disableFlattening: 'false'
+        }
```
Spreading is unnecessary here. There's no conditional wrapping the object. Move props above.

rart commented on ui/app/src/services/content.ts:
```diff
@@ -562,6 +646,69 @@ export function insertComponent(
+        }
+      });
+
+      let fieldNode = doc.querySelector(`:scope > ${fieldId}`);
```
Need to update to use extractNode, etc. Refactor the code duplication with `insertComponent`.

rart commented on ui/app/src/state/actions/preview.ts:
```diff
@@ -17,12 +17,12 @@
 import { StandardAction } from '../../models/StandardAction';
 import ContentType from '../../models/ContentType';
-import ContentInstance from '../../models/ContentInstance';
+import ContentInstance, { SearchContentInstance } from '../../models/ContentInstance';
```
Move "SearchContentInstance" to Search.

Rename to ContentInstancePage.

rart commented on ui/app/src/utils/hooks.ts:
```diff
@@ -155,3 +157,15 @@ export function useStateResource<ReturnType = any, SourceType = GlobalState>(
   useEffect(componentDidMount, []);
 }
+
+export function useDebouncedInput(observer: (keywords: string) => any) {
```
- add debounceTime param
- default it to 250 (ms)
- use it for debounce
- add to effect deps

rart commented on ui/app/src/utils/string.ts:
```diff
@@ -101,3 +93,17 @@ export function dataUriToBlob(dataURI: string) {
 }
+
+export function objectIdFromPath(path: string) {
+  return path.substr(path.lastIndexOf('/') + 1).replace(/\.xml/, '')
```
;

rart commented on ui/guest/src/classes/ContentController.js:
```diff
@@ -864,6 +890,19 @@ function reducer(lookupTable, model) {
 
 }
 
+function getResult(model, fieldId, index) {
```
Make sure to test with the different level possibilities. 

- Collections at first level
- Nested collections

rart commented on ui/app/src/modules/Preview/Tools/AssetsPanel.tsx:
```diff
@@ -260,15 +254,11 @@ export default function AssetsPanel() {
-    });
-    return () => subscription.unsubscribe();
-  }, [dispatch, onSearch$]);
+  const onSearch = useMemo(() => (
```
`useCallback` instead of `useMemo`

aachavarria committed:
PR Feedback

aachavarria committed:
Merge branch '2019-11-ui-restructure' of https://github.com/craftercms/studio-ui into feature/3756

 Conflicts:
	ui/app/src/services/content.ts

aachavarria committed:
AssetsPanel: fix dependencies

aachavarria committed:
styles.js: adding styles for spinner;
IceRegistry.js: fixing validation to support string

aachavarria committed:
previewConcierge: adding case COMPONENT_INSTANCE_HTML_REQUEST;
content.ts: adding getComponentInstanceHTML(), adding performMutation to insertMethods, refactor insert methods;
contentController: adding parentModelId;
Guest.js, guestProxy: insert changes to support nested;

aachavarria committed:
revert .env change

aachavarria committed:
camelCase

rart commented on ui/app/src/modules/Preview/PreviewConcierge.tsx:
```diff
@@ -267,6 +301,14 @@ export function PreviewConcierge(props: any) {
+              type: COMPONENT_INSTANCE_HTML_RESPONSE,
+              payload: htmlString
+            });
+          });
```
ID?

rart commented on ui/app/src/modules/Preview/Tools/AssetsPanel.tsx:
```diff
@@ -189,7 +184,8 @@ export default function AssetsPanel() {
   });
 
-  const onDragDrop = (e) => {
+  //useCallback(() => {},[])
```
x

rart commented on ui/app/src/modules/Preview/Tools/BrowseComponentsPanel.tsx:
```diff
@@ -0,0 +1,321 @@
+export default function BrowseComponentsPanel() {
+
+  const classes = useStyles({});
+  //const onSearch$ = useMemo(() => new Subject<string>(), []);
```
x

rart commented on ui/app/src/services/content.ts:
```diff
@@ -49,6 +58,89 @@ export function getDOM(site: string, path: string): Observable<XMLDocument> {
+  );
+}
+
+//sent an id compuesto?
```
x

rart commented on ui/app/src/services/content.ts:
```diff
@@ -31,13 +31,22 @@ import {
+export function getComponentInstanceHTML(path: string): Observable<string> {
+  return fromPromise(fetch(`/crafter-controller/component.html?path=${path}`)).pipe(
+    switchMap(response => response.text()),
+    map(response => response.replace(/<\!--.*?-->/g, ''))
```
x

aachavarria committed:
assetPanel fix

rart commented on ui/app/src/services/content.ts:
```diff
@@ -49,6 +58,89 @@ export function getDOM(site: string, path: string): Observable<XMLDocument> {
+  }
+}
+
+function parseContentXML(doc: XMLDocument, path: string = null, contentTypesLookup: LookupTable<ContentType>, lookup: LookupTable<ContentInstance>): LookupTable<ContentInstance> {
```
rename `lookup` to `instanceLookup`

rart commented on ui/app/src/services/content.ts:
```diff
@@ -49,6 +58,89 @@ export function getDOM(site: string, path: string): Observable<XMLDocument> {
+}
+
+//sent an id compuesto?
+function parseElementByContentType(element: Element, field: ContentTypeField, contentTypesLookup: LookupTable<ContentType>, lookup: LookupTable<ContentInstance>) {
```
rename `lookup` to `instanceLookup`

rart commented on ui/app/src/services/content.ts:
```diff
@@ -49,6 +58,89 @@ export function getDOM(site: string, path: string): Observable<XMLDocument> {
+  return lookup;
+}
+
+const skippableList = [
```
systemPropsList

rart commented on ui/app/src/services/content.ts:
```diff
@@ -533,32 +620,120 @@ export function insertComponent(
+  //     );
+  //
+  //   })
+  // );
```
x

rart commented on ui/app/src/services/content.ts:
```diff
@@ -852,6 +1036,30 @@ function updateModifiedDateElement(doc: XMLDocument) {
+  return `${pathBase}/${id}.xml`;
+}
+
+function insertItemOnDoc(doc: XMLDocument, fieldId: string, targetIndex: string | number, newItem: Node) {
```
Specify return type

Rename to `insertColIectionItem`

make sure it works with nested

aachavarria replied:
yep it works

rart commented on ui/app/src/services/content.ts:
```diff
@@ -724,30 +899,39 @@ export function deleteItem(
+    },
+    {
+      'Content-Type': 'application/json'
     }
```
use `postJSON` from `util/ajax`.

Or if it doesn't cut it, at least use `CONTENT_TYPE_JSON` from same util file.

rart commented on ui/app/src/state/reducers/preview.ts:
```diff
@@ -274,7 +295,44 @@ const reducer = createReducer<GlobalState['preview']>({
+    selectedTool: 'craftercms.ice.browseComponents',
+    components: { ...state.components, contentTypeFilter: payload }
+  }),
 });
```
Page change + search bug (reset page number on search)

aachavarria committed:
adding and using getText();
adding date to COMPONENT_INSTANCE_HTML_REQUEST as id;

rart commented on ui/app/src/utils/content.ts:
```diff
@@ -30,7 +30,14 @@ export function isEditableAsset(path: string) {
+  var txt = document.createElement('textarea');
+  txt.innerHTML = html;
+  return txt.value;
+}
```
Add param and return types.
Move to string utils.

Why not use actual decode functions?

aachavarria replied:
which one?

rart commented on ui/guest/src/components/Guest.js:
```diff
@@ -490,7 +552,7 @@ export function Guest(props) {
       });
 
-      // scrollToReceptacle(validatedReceptacles);
+      scrollToReceptacle(validatedReceptacles);
```
revert (put the comment back)

rart commented on ui/guest/src/components/GuestProxy.js:
```diff
@@ -250,46 +282,63 @@ export function GuestProxy(props) {
+            $spinner.remove();
+            insertElement($component, $daddy, targetIndex);
+            updateElementRegistrations(Array.from($daddy.children()), 'insert', targetIndex);
+            //TODO:update register of all fields on the new
```
What's needed to complete this?

rart commented on ui/guest/src/styles.js:
```diff
@@ -156,3 +171,41 @@ export function appendStyleSheet() {
+//     stroke-dasharray: 90, 150;
+//   stroke-dashoffset: -124;
+// }
+// }
```
x

aachavarria committed:
renaming lookup to instanceLookup

aachavarria committed:
renaming

aachavarria committed:
content.ts, string.ts: moving decodeHTML();
guest.js: commenting scrollToReceptacle

aachavarria committed:
removing comment

aachavarria committed:
registering new component elements

aachavarria committed:
using postJSON

rart commented on ui/app/src/services/content.ts:
```diff
@@ -477,25 +573,20 @@ export function insertComponent(
   modelId: string,
   fieldId: string,
-  targetIndex: number,
+  targetIndex: string,
```
string | number

rart commented on ui/guest/src/components/GuestProxy.js:
```diff
@@ -250,46 +282,63 @@ export function GuestProxy(props) {
+          const date = Date.now();
+
+          message$.pipe(
+            filter((e) => (e.data?.type === COMPONENT_INSTANCE_HTML_RESPONSE) && (e.data?.payload.date === date)),
```
`date` => `id`

aachavarria committed:
PR Feedback

aachavarria committed:
changing date to id

rart committed:
Update GuestProxy.js

rart commented on ui/guest/src/components/GuestProxy.js:
```diff
@@ -250,46 +282,63 @@ export function GuestProxy(props) {
+            $component.attr('data-craftercms-field-id', fieldId);
+            $spinner.remove();
+            insertElement($component, $daddy, targetIndex);
+            updateElementRegistrations(Array.from($daddy.children()), 'insert', targetIndex);
```
Consider nested cases.

Index may be `0.1`. You'd need to send it the last piece (possibly as an int).

Check other usages of `updateElementRegistrations`.

rart replied:
remember you can also use `$().children().toArray()`

aachavarria committed:
supporting nested on updateElementRegistrations()

aachavarria committed:
Merge branch '2019-11-ui-restructure' of https://github.com/craftercms/studio-ui into feature/3756

 Conflicts:
	ui/app/src/models/GlobalState.ts
	ui/app/src/modules/Preview/PreviewConcierge.tsx
	ui/app/src/state/actions/preview.ts
	ui/app/src/state/reducers/preview.ts
